grayscale.py

```diff
@@ -1,764 +1,549 @@
-        # well as any timing offsets introduced by the Python code.
-        # 0xd3,x    Set display offset. Since rows are scanned in reverse, the
-        #           calculation must work backwards from the last controller row.
+        self._preFrameCmds = bytearray([0xa8,0, 0xd3,52])
+        # Once the frame has been loaded into the display controller's
+        # GDRAM, we set the controller to output 57 rows, and then delay
+        # for the nominal time for 48 rows to be output.
+        # Considering the 17 row 'buffer space' after the real 40 rows,
+        # that puts us around halfway between the end of the display, and
+        # the row at which it would wrap around.
+        # By having 8.5 rows either side of the nominal timing, we can
+        # absorb any variation in the frequency of the display controller's
+        # RC oscillator as well as any timing offsets introduced by the
+        # Python code.
+        # 0xd3,x    Set display offset. Since rows are scanned in reverse,
+        #           the calculation must work backwards from the last
+        #.          controller row.
         # 0xa8,57-1 Set multiplex ratio to 57
-        self.post_frame_cmds = bytearray([0xd3,40+(64-57), 0xa8,57-1])
+        self._postFrameCmds = bytearray([0xd3,40+(64-57), 0xa8,57-1])
 
-        # We enhance the greys by modulating the contrast.
-        # Use setting from thumby.cfg
-        brightnessSetting=2
+        # Brightness modulation.
+        # We enhance the greys by modulating the contrast,
+        # limited to brightness setting from thumby.cfg
+        brightness = 127
         try:
             with open("thumby.cfg", "r") as fh:
-                conf = fh.read().split(',')
-            for k in range(len(conf)):
-                if(conf[k] == "brightness"):
-                    brightnessSetting = int(conf[k+1])
-        except OSError:
+                _, _, conf = fh.read().partition("brightness,")
+                b = int(conf.split(',')[0])
+                # Set to the relevant brightness level
+                brightness = 0 if b==0 else 56 if b==1 else 127
+        except (OSError, ValueError):
             pass
-        # but with a set of contrast values spanning the entire range provided by the controller
-        brightnessVals = [0,56,127]
-        brightnessVal = brightnessVals[brightnessSetting]
-        # 0x81,<val>        Set Bank0 contrast value to <val>
-        self.post_frame_adj = [bytearray([0x81,brightnessVal>>5]), bytearray([0x81,brightnessVal]), bytearray([0x81,(brightnessVal << 1) + 1])]
+        # 0x81,<val>    Set Bank0 contrast value to <val>
+        self._postFrameAdj = array('O', [
+            bytearray([0x81,brightness>>5]),
+            bytearray([0x81,brightness]),
+            bytearray([0x81,(brightness << 1) + 1])])
 
-        # It's better to avoid using regular variables for thread sychronisation.
+        # It's safer to avoid using regular variables for shared thread data.
```

Comment:
I would love to understand this then. The comment just made me wonder why its better.

Reply:
I changed it back now anyway.

Reply:
Sorry, I think 'safer' is probably more correct here. It's not safer against crashes (that should be more of a v1.18 issue), but it will be safer against synchronisation issues relating to object reference updates or the atomicity of updating an int in a 32-bit int array. That's how it's 'better', but really 'safer' still applies.